bug: message won't be sent over from node1 to node2 with sharded topic subscription

**Description**
During TestStaticShardingLimits execution 1024 pubsub topics were subscribed on relay at node1 and node2.  Relay instance1 was used to publish one message on randomly selected pubsub topic out of those 1024 subscribed. Subscription on relay at node2 is used to check the message was received. It looks message was not received. I have added relay.IsSubscribed() check on both relays, and this check is passing. 

Similar scenario is working at TestWakuRelayStaticSharding with the difference message is sent from relay2 to relay1.

**To Reproduce**
1) Checkout https://github.com/waku-org/go-waku/pull/1060/commits/e1210c7ce3b5d3eb3d3ae1284855d465cfe4e1df
2) Apply patch from https://github.com/waku-org/go-waku/pull/1084
3) cd go-waku/waku/v2/node
4) go test -run TestStaticShardingLimits

**Expected behavior**
Message could be received on node2 after publishing on node1.

**Log**
```
--- FAIL: TestStaticShardingLimits (33.97s)
    utils.go:402:
        	Error Trace:	/Users/roman/sources/waku-org/go-waku/tests/utils.go:402
        	            				/usr/local/go/src/runtime/asm_arm64.s:1172
        	Error:      	Message timeout
        	Test:       	TestStaticShardingLimits
FAIL
```

I modified the code slightly:
```
diff --git a/waku/v2/node/wakunode2_test.go b/waku/v2/node/wakunode2_test.go
index eee80e80..7cb94d38 100644
--- a/waku/v2/node/wakunode2_test.go
+++ b/waku/v2/node/wakunode2_test.go
@@ -4,7 +4,6 @@ import (
 	"bytes"
 	"context"
 	"fmt"
-	wenr "github.com/waku-org/go-waku/waku/v2/protocol/enr"
 	"math/big"
 	"math/rand"
 	"net"
@@ -12,6 +11,8 @@ import (
 	"testing"
 	"time"
 
+	wenr "github.com/waku-org/go-waku/waku/v2/protocol/enr"
+
 	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/p2p/enode"
 	"github.com/prometheus/client_golang/prometheus"
@@ -418,7 +419,7 @@ func TestStaticShardingMultipleTopics(t *testing.T) {
 }
 
 func TestStaticShardingLimits(t *testing.T) {
-	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 1000*time.Second)
 	defer cancel()
 
 	testClusterID := uint16(21)
@@ -507,6 +508,8 @@ func TestStaticShardingLimits(t *testing.T) {
 	require.True(t, r1.IsSubscribed(shardedPubSubTopics[randomShard]))
 	require.True(t, r2.IsSubscribed(shardedPubSubTopics[randomShard]))
 
+	time.Sleep(120 * time.Second)
+
 	// Publish on node1
 	_, err = r1.Publish(ctx, msg1, relay.WithPubSubTopic(shardedPubSubTopics[randomShard]))
 	require.NoError(t, err)
```

And noticed that with this sleep, the test case is succesful:
```
2024-04-10T09:57:14.708-0400    INFO    gowaku.node2.peer-manager       peermanager/peer_manager.go:441 peer already found in peerstore, but re-adding it as ENR sequence is higher than locally stored   {"peer": "16Uiu2HAmCFRE1yenDPuRK91gMTbJm9iEuMWFZwG5Qmxc9Fk7n9vZ", "newENRSeq": 1712757407949, "storedENRSeq": 1712757407346}
2024-04-10T09:57:14.709-0400    INFO    gowaku.node2.peer-manager       peermanager/peer_manager.go:474 adding peer to peerstore        {"peer": "16Uiu2HAmCFRE1yenDPuRK91gMTbJm9iEuMWFZwG5Qmxc9Fk7n9vZ"}
2024-04-10T09:58:00.945-0400    INFO    gowaku.node2.connection-notifier        node/connectedness.go:58        peer connected  {"peer": "16Uiu2HAmCFRE1yenDPuRK91gMTbJm9iEuMWFZwG5Qmxc9Fk7n9vZ", "direction": "Outbound"}
2024-04-10T09:58:00.945-0400    INFO    gowaku.node2.connection-notifier        node/connectedness.go:58        peer connected  {"peer": "16Uiu2HAm5uw5LNkbVeJizxDs5PEszEHcSYHgQHBgpBYmSEjNEHb8", "direction": "Inbound"}
2024-04-10T09:59:16.642-0400    WARN    pubsub  go-libp2p-pubsub@v0.10.0/tag_tracer.go:157      error bumping delivery tag: no decaying tag registered for topic /waku/2/rs/21/920
2024-04-10T09:59:17.643-0400    INFO    gowaku  tests/utils.go:400      Received        {"msg": "payload:\"test message\" content_topic:\"/test/2/my-app/sharded\" timestamp:1712757436641541548"}
```
However, do note that it took 45s for the peer to be discovered and connected (check the delta between line 2 and 3 of the logs). Not sure what is the correct behavior, tho. I imagine that the time it took depends on how long running the connectivity loop takes between iterations.

cc: @chaitanyaprem

--- a/waku/v2/peermanager/peer_connector.go
+++ b/waku/v2/peermanager/peer_connector.go
@@ -193,7 +193,6 @@ func (c *PeerConnectionStrategy) canDialPeer(pi peer.AddrInfo) bool {
 	c.mux.Lock()
 	defer c.mux.Unlock()
 	val, ok := c.cache.Get(pi.ID)
-	var cachedPeer *connCacheData
 	if ok {
 		tv := val.(*connCacheData)
 		now := time.Now()
@@ -204,15 +203,25 @@ func (c *PeerConnectionStrategy) canDialPeer(pi peer.AddrInfo) bool {
 		}
 		c.logger.Debug("Proceeding with connecting to peer",
 			zap.Time("currentTime", now), zap.Time("nextTry", tv.nextTry))
-		tv.nextTry = now.Add(tv.strat.Delay())
+	}
+	return true
+}
+
+func (c *PeerConnectionStrategy) addConnectionBackoff(peerID peer.ID) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	val, ok := c.cache.Get(peerID)
+	var cachedPeer *connCacheData
+	if ok {
+		tv := val.(*connCacheData)
+		tv.nextTry = time.Now().Add(tv.strat.Delay())
 	} else {
 		cachedPeer = &connCacheData{strat: c.backoff()}
 		cachedPeer.nextTry = time.Now().Add(cachedPeer.strat.Delay())
 		c.logger.Debug("Initializing connectionCache for peer ",
-			logging.HostID("peerID", pi.ID), zap.Time("until", cachedPeer.nextTry))
-		c.cache.Add(pi.ID, cachedPeer)
+			logging.HostID("peerID", peerID), zap.Time("until", cachedPeer.nextTry))
+		c.cache.Add(peerID, cachedPeer)
 	}
-	return true
 }
 
 func (c *PeerConnectionStrategy) dialPeers() {
@@ -255,6 +264,7 @@ func (c *PeerConnectionStrategy) dialPeer(pi peer.AddrInfo, sem chan struct{}) {
 	defer cancel()
 	err := c.host.Connect(ctx, pi)
 	if err != nil && !errors.Is(err, context.Canceled) {
+		c.addConnectionBackoff(pi.ID)
 		c.host.Peerstore().(wps.WakuPeerstore).AddConnFailure(pi)
 		c.logger.Warn("connecting to peer", logging.HostID("peerID", pi.ID), zap.Error(err))
 	}

